pallets/drop3/src/lib.rs

```diff
@@ -0,0 +1,508 @@
+//!     pool multiple times, or close the pool to remove it from the map storage,
+//!     where the remaining amount will be unreserved.
+//!  4. when the pool is actively running, the owner can send the reward to any other user,
+//!     the amount will be deducted directly from reserved balance.
+//!
+//! Some notes:
+//!  - the admin account can only be set by SetAdminOrigin, which will be bound at runtime.
+//!  - a user can propose/own multiple reward pools.
+//!  - the events and errors are relatively in verbose level.
+//!  - about the usage of reversed balance: see the checks around `slash_reserved` and
+//!    `repatriated_reversed`, theoretically the desired amount could always be slashed/moved, as
+//!    `total` amount is already reserved upon creation of a reward pool. However, we are playing it
+//!    safe and we make sure the emitted event always contains the actual amount, since we don't
+//!    know if user's balance is unexpectedly unreserved somewhere outside this pallet.
+
+#![allow(clippy::type_complexity)]
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+pub mod weights;
+
+use frame_support::{
+	pallet_prelude::*,
+	traits::{BalanceStatus, ReservableCurrency, StorageVersion},
+};
+use frame_system::pallet_prelude::*;
+pub use pallet::*;
+use sp_runtime::{
+	traits::{AtLeast32BitUnsigned, Bounded, Saturating, Zero},
+	Percent,
+};
+
+use scale_info::TypeInfo;
+use sp_std::vec::Vec;
+pub use weights::WeightInfo;
+
+/// a single reward pool
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+pub struct RewardPool<PoolId, BoundedString, AccountId, Balance, BlockNumber> {
+	// unique pool id
+	id: PoolId,
+	// a bounded string whose length is limited by `MaximumNameLength`
+	name: BoundedString,
+	// account id of the pool owner
+	owner: AccountId,
+	// total amount of token that will be reserved upon creation
+	total: Balance,
+	// remaining amount of token that can be sent as reward for this pool
+	remain: Balance,
+	// block height where the pool was created
+	create_at: BlockNumber,
+	// start block height which is defined when proposing,
+	// calling `send_reward` prior to this height would fail
+	start_at: BlockNumber,
+	// end block height which is defined when proposing,
+	// calling `send_reward` after this height would fail
+	end_at: BlockNumber,
+	// if the pool is started
+	started: bool,
+	// if the pool is approved
+	approved: bool,
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	/// The current storage version.
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: pallet_balances::Config + frame_system::Config {
```

Comment:
maybe we could add a "type Currency: ReservableCurrency<Self::AccountId>" instead of balance pallet;
As in future we might use multi token or other balance pallet instead?

Reply:
I see both usage in substrate, but feel a single type make more sense in our condition.

Reply:
> maybe we could add a "type Currency: ReservableCurrencySelf::AccountId" instead of balance pallet; As in future we might use multi token or other balance pallet instead?

> Agree with Minqi, probably `Currency<Self::AccountId> + ReservableCurrency<Self::AccountId>` is more commonly used.

Good point, I'll think twice and look forward to changing it

Reply:
Now adjusted.